Body read errors should return Httpincomplete instead of Error

An upstream request appears to have timed out, and the error returned to the SDK was a generic Error and not HttpIncomplete.

```
2023-07-18T20:29:05.099307Z TRACE wiggle abi{module="fastly_http_body" function="read"}: result=Err(HyperError(hyper::Error(Body, Custom { kind: UnexpectedEof, error: "unexpected EOF during chunk size line" })))
2023-07-18T20:29:05.099330Z ERROR wiggle abi{module="fastly_http_body" function="read"}: Hostcall yielded an error: error reading a body from connection: unexpected EOF during chunk size line
    ReadByte: Fastly error: Error
    Code = 502; want 200
    Body = "11223344Fastly error: Error\n"; want "112233445566778899001122\n\n"
--- FAIL: TestByteRepeater (15.09s)
```

--- a/lib/src/error.rs
+++ b/lib/src/error.rs
@@ -1,5 +1,7 @@
 //! Error types.
 
+use std::error::Error as StdError;
+use std::io;
 use {crate::wiggle_abi::types::FastlyStatus, url::Url, wiggle::GuestError};
 
 #[derive(Debug, thiserror::Error)]
@@ -153,6 +155,14 @@ impl Error {
             Error::HyperError(e) if e.is_parse() => FastlyStatus::Httpinvalid,
             Error::HyperError(e) if e.is_user() => FastlyStatus::Httpuser,
             Error::HyperError(e) if e.is_incomplete_message() => FastlyStatus::Httpincomplete,
+            Error::HyperError(e)
+                if e.source()
+                    .and_then(|e| e.downcast_ref::<io::Error>())
+                    .map(|ioe| ioe.kind())
+                    == Some(io::ErrorKind::UnexpectedEof) =>
+            {
+                FastlyStatus::Httpincomplete
+            }
             Error::HyperError(_) => FastlyStatus::Error,
             // Destructuring a GuestError is recursive, so we use a helper function:
             Error::GuestError(e) => Self::guest_error_fastly_status(e),

